jit/core/jit_cleaner.py

```diff
@@ -0,0 +1,160 @@
+    if len(policy.get("bindings", [])) == 0:
+        jit_logger.error("project binding empty. ")
+        return
+    binding_count = 0
+    for b in policy["bindings"]:
+        if (
+          role == b["role"] and
+          member in b["members"] and
+          expression == b.get("condition", {}).get("expression", "")
+        ):
+            b["members"].remove(member)
+            binding_count += 1
+            continue
+
+    policy["bindings"][:] = [b for b in policy["bindings"] if b["members"]]
+    # if "members" in binding and member in binding["members"]:
+    #     binding["members"].remove(member)
+    # jit_logger.info(binding)
+
+    service.projects().setIamPolicy(
+      resource=target_project,
+      body={"policy": policy}
+    ).execute()
+    jit_logger.info("%d binding has been removed.", binding_count)
+
+
+def process_pubsub_msg(conf, received_message):
+    jit_logger.info(f"Received: {received_message.message.data}.")
+    jit_logger.info(f"attributes: {received_message.message.attributes}")
+    origin = received_message.message.attributes.get("origin", "")
+    # for invalid origin, we should directly return and make the run_jit_cleaner
+    # ack the message.
+    if not origin or constant.MessageOrigin.from_str(
+      origin) != constant.MessageOrigin.BINDING:
+        jit_logger.error(
+          f"origin invalid: origin={origin}, msg data = {received_message.message.data}.")
+        return
+
+        # if any of these process invalid, it may raise exception,
+    # the exception will be captured in run_jit_cleaner level
+    pubsub_msg_dict = json.loads(received_message.message.data)
+    target_project_id = pubsub_msg_dict["project_id"]
+    condition = pubsub_msg_dict["conditions"]
+    expression = condition["expression"]
+    member = "user:{user}".format(user=pubsub_msg_dict["user"])
+    role = pubsub_msg_dict["role"]
+
+    start = expression.get("start", "1900-01-01T00:00:00.00000Z")
+    # start_datetime = datetime.fromisoformat(start)
+    start_datetime = datetime.strptime(start, "%Y-%m-%dT%H:%M:%S.%fZ")
+    end = expression.get("end", "1900-01-01T00:00:00.00000Z")
+    # end_datetime = datetime.fromisoformat(end)
+    end_datetime = datetime.strptime(end, "%Y-%m-%dT%H:%M:%S.%fZ")
+
+    # filter by date.
+    # 1,2,3,4
+    #
+    if end_datetime < datetime.utcnow():
+        jit_logger.info(f"access expired: {received_message.message}")
+        modify_policy_remove_member(conf, target_project_id, role, member,
+                                    start, end)
+    else:
+        publisher = pubsub_v1.PublisherClient()
+        future = publisher.publish(
+          conf.pubsub_topic_name,
+          received_message.message.data,
+          origin=origin
+        )
+        jit_logger.info("putting back to queue. msg_id=%s", future.result())
+
+
+def run_jit_cleaner(conf: config.JitConfig):
+    """
+    jit_cleaner fetches message from pubsub with max=NUM_MESSAGES
```

Comment:
So I see a small issue here. If max NUM_MESSAGES is less than the messages on the queue then how do you ensure that you've processes all expiration messages?

Reply:
as previously discussed, we need to process batch by batch. if we consume all the message it may possibly cause timeout. we can either ignore the timeout first or ignore the NUM_MESSAGES first (cus we can set it high and run it more frequent).

changing it to consume all message, if you prefer to do so.